### PR TITLE
Use .alt_entry/.private_extern on Apple platforms

### PR DESCRIPTION
**Pull request by filipnavara**
# Background

Unlike ELF which can support pretty much arbitrary number of sections, the Mach-O object files are limited to ~255 sections. Apple uses the `.subsections_via_symbols` linking approach to circumvent this limitation. In this model the external symbols are used to slice the section into subsections, or atoms, which are then used as the linking unit (for purpose of dead stripping, reordering, etc.).

`.alt_entry` is an escape hatch to introduce symbols that don't divide a (sub)section into multiple atoms. It can be combined with `.global` or `.private_extern` to create symbols pointing into the middle of a method or data while making sure that the linker doesn't try to slice the method/data and place it into two different locations in the output.

`.private_extern` is similar to combination of symbol marked with `.global` and `.hidden` on ELF platforms. It's used to mark symbols with limited visibility. When the file is fed to the static linker, it clears the `N_EXT` bit for each symbol with the N_PEXT bit set. *The ld option `-keep_private_externs` turns off this behavior.*

# Problem

NativeAOT and CoreCLR have assembly files that need to reference instructions in the middle of the method. This is used for marking places recognized by the `SIGSEGV` handler as valid instructions that are expected to produce a signal which can transformed to `NullReferenceException`. CoreCLR also uses it in parts of code that are patchable at runtime (applies only to macOS x64 for the purpose of this PR).

Additionally, NativeAOT has a concept of dehydrated data that are unpacked at runtime into the `.hydrated` section. The dehydrated data use relative offsets and rely on the linker keeping the layout of the `.hydrated` section intact.

`.subsections_via_symbols` was emitted by Clang for over a decade now and the new Apple linker introduced in Xcode 15 no longer correctly supports object files without it. The old linker is scheduled to be removed in Xcode 16 which is currently in beta. Additionally, Xcode 16 is more strict at enforcing the rules for `.global` labels inside assembly files. While the current Xcode 16 betas have additional issues (upstream LLVM links: https://github.com/llvm/llvm-project/issues/82261, https://github.com/llvm/llvm-project/issues/97116), it is prudent to change the code to emit `.alt_entry` symbols instead of depending on undefined linker behavior. 

# Solution

This PR is changing the code to correctly use `.alt_entry`/`.private_extern` and compile with Xcode 14/15.

It's best reviewed commit-by-commit:
- [13e6013](https://github.com/dotnet/runtime/pull/106224/commits/13e6013a4f2b7ac9692d60e653b77bc98d56aef2) updates macros used in NativeAOT assembly code to use `.alt_entry` + `.private_extern` instead of `.global` to mark code location labels.
- [5fe3a06](https://github.com/dotnet/runtime/pull/106224/commits/5fe3a0655a0ddc5c71f93bb0d2b3608db37949b7) updates the ILC compiler to emit `N_ALT_ENTRY` flag for any symbols pointing to the middle of the node.
- [ea6ab79](https://github.com/dotnet/runtime/pull/106224/commits/ea6ab79859ffe00e81d5eed5bf1896904d676584) updates the ILC compiler to treat the whole `.hydrated` section as single unbreakable atom for the linker. It also updates the Mach-O emitter to emit the `N_PEXT` (`.private_extern`) flag to match the ELF behavior.

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs [hunk 1]
```diff
@@ -483,8 +504,10 @@ protected internal override void UpdateSectionAlignment(int sectionIndex, int al
                     Name = name,
                     Section = section,
                     Value = section.VirtualAddress + (ulong)definition.Value,
-                    Descriptor = N_NO_DEAD_STRIP,
-                    Type = N_SECT | N_EXT,
+                    Descriptor =
+                        definition.AltEntry || definition.SectionIndex == _hydrationTargetSectionIndex ?
+                        N_ALT_ENTRY : N_NO_DEAD_STRIP,
```
Comment (filipnavara) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs:509]: Strictly speaking the `N_NO_DEAD_STRIP` should no longer be required with the changes in this PR. I left it in out of abundance of caution and to minimize the chance of something going wrong.

**Comment by filipnavara** [comment]
Temporarily switched back to draft while I try to validate what is happening with the new linker (Xcode 15/16). While we currently disable it I want to ensure that this doesn't introduce any additional regression that we would have to deal with. Seems like the `.hydrated` section may not be linker correctly with it.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs (hunk 1 above)
Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs]: Turns out, caution was warranted. New Xcode linker still messes up the `.hydrated` section even if it's completely composed of `N_ALT_ENTRY` symbols. I changed it to `N_ALT_ENTRY | N_NO_DEAD_STRIP` and I will probably follow up with a bug report to Apple.

**Comment by filipnavara** [comment]
Re-tested with the following scenarios:
- Xcode 15.4 running the whole build, building and running NativeAOT smoke tests
- Xcode 15.4 used for building the runtime and libs; Xcode 16 beta 5 and new linker used for building the NativeAOT smoke tests. I switched the active Xcode after building the runtime and commented out these lines in artifacts/bin/coreclr/osx.arm64.Release/build/Microsoft.NETCore.Native.Unix.targets: https://github.com/dotnet/runtime/blob/fa33eddf3533dac554e959da92f54104c0568c1e/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L256-L258

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs (hunk 1 above)
Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs]: I found the pattern, will fix ILC to avoid producing it:

test_data.S:
```asm
.subsections_via_symbols
.section __DATA,hydrated
.global .hydrated // <-- symbol we generate at beginning of .hydrated section
.hydrated:
lsection6: // <-- local symbol for beginning of section
.alt_entry _bar_data
.private_extern _foo_data
_foo_data:
  .dword 0
.alt_entry _bar_data
.private_extern _bar_data
_bar_data:
  .dword 1
```

test_code.S:
```
extern char *foo_data;
int main() { return (int)(long)&foo_data; }
```

Build with `clang test_data.S test_code.c -o ./test -Wl,-dead_strip [-Wl,-ld_classic]`. Check output with `nm -nm ./test`

For new linker we get:
```
0000000100000000 (__TEXT,__text) [referenced dynamically] external __mh_execute_header
0000000100003f8c (__TEXT,__text) external _main
0000000100004000 (__DATA,hydrated) external .hydrated
0000000100004000 (__DATA,hydrated) non-external (was a private external) _foo_data
```

For old linker we get:
```
0000000100000000 (__TEXT,__text) [referenced dynamically] external __mh_execute_header
0000000100003f9c (__TEXT,__text) external _main
0000000100004000 (__DATA,hydrated) external .hydrated
0000000100004000 (__DATA,hydrated) non-external (was a private external) _foo_data
0000000100004008 (__DATA,hydrated) non-external (was a private external) _bar_data
```

This only happens if we produce the local section symbol.

Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs]: Filed as Apple bug FB14743667. The workaround in ILC is orthogonal to the changes in this PR, so I will likely submit it separately after I do sufficient testing. Keeping `N_NO_DEAD_STRIP` is sufficient workaround for now.

Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs]: Funnily enough, if I rename `lsectionXX` to `ltmpXX` (ie. the same symbol prefix as emitted for LLVM temporary variables) it works. That seems to be cheap enough workaround.

**Comment by filipnavara** [comment]
> Does this work if dehydration is turned off?

Yes

> I don't fully understand why the hydrated section is special. Is it because we happen to place everything that cannot be broken up into that section right now (when dehydration is turned on, that is)?

It's special because the dehydrated data indexes the `hydrated` section using relative offsets embedded in the dehydrated data that are not understood by the native linker.

In the atom linking model the native linker is free to reorder atoms (subsections), or discard them if they are unreferenced (from the linker's point of view, ie. relocation targets). Thus the whole "hydrated" section is by default considered a bunch of atoms that can be reordered or removed. If that happens the offsets in the dehydrated data no longer match the symbols and relocations that point to them.

The idea is to turn the whole section into one single non-breakable and non-relocatable atom for the native linker. That's accomplished by making the whole section start with the ".hydrated" symbol, which is the start of the atom, and then marking any other symbol in the section as "alternative entry" symbol. "alternative entry" symbols don't break up atoms (*).

We previously started to mark all the symbols with "no dead strip" marker. That guarantees that the symbol is preserved in the final linked image. It doesn't guarantee that it's present in the original order which is a stronger guarantee that this PR is trying to enforce with documented behavior instead of relying on undefined behavior.

(*) Unless you hit a linker bug, like with the new Xcode 16 linker

**Comment by MichalStrehovsky** [comment]
> It's special because the dehydrated data indexes the hydrated section using relative offsets embedded in the dehydrated data that are not understood by the native linker.

Don't we do the same thing when dehydration is turned off? E.g. the fact that MethodTable symbols are prefixed by GCDesc, or non-GC static bases are prefixed by the class constructor context. If the linker decides to shuffle these, bad things will happen. I would expect we do the same thing for all sections we emit. The compiler doesn't expect linker messing with section contents.

**Comment by filipnavara** [comment]
> Don't we do the same thing when dehydration is turned off? E.g. the fact that MethodTable symbols are prefixed by GCDesc, or non-GC static bases are prefixed by the class constructor context. If the linker decides to shuffle these, bad things will happen. I would expect we do the same thing for all sections we emit. The compiler doesn't expect linker messing with section contents.

That largely depends on which symbols are produced in the object file and whether they are referenced or not. As far as I can tell, `EETypeNode` calls `ObjectDataBuilder.AddSymbol` only once. If you have single symbol for a node, you are creating a single atom for the native linker.

For places which call `ObjectDataBuilder.AddSymbol` more than once it would result in more atoms in the native linker (*). They are currently all marked with `N_NO_DEAD_STRIP`, so the linker will not remove them. It will also very likely not reorder them but that's not guaranteed without `N_ALT_ENTRY` merging the symbols into single atom. Commit https://github.com/dotnet/runtime/pull/106224/commits/5fe3a0655a0ddc5c71f93bb0d2b3608db37949b7 in this PR does exactly that.

(*) ...or rather it may result in more atoms in the linker. The heuristics for code also look at unwinding information, so I am more concerned with DATA.

**Comment by MichalStrehovsky** [comment]
> That largely depends on which symbols are produced in the object file and whether they are referenced or not. As far as I can tell, EETypeNode calls ObjectDataBuilder.AddSymbol only once. If you have single symbol for a node, you are creating a single atom for the native linker.

EETypeNode sets ISymbolDefinition.Offset to the GCDesc size:

https://github.com/dotnet/runtime/blob/15e96faf4558b017ea8df1dc28d9b2169f0badc0/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs#L247-L248

So the symbol is only defined GCDesc-size-bytes into the ObjectData blob. Would the data still be emitted correctly for this?

If I understand this correctly, we'd need object writer to emit a (made up?) symbol at the beginning of object data, (make sure it's referenced somehow?) and generate the actual MethodTable symbol as alt-entry.

**Comment by filipnavara** [comment]
> So the symbol is only defined GCDesc-size-bytes into the ObjectData blob. Would the data still be emitted correctly for this?
> 
> If I understand this correctly, we'd need object writer to emit a (made up?) symbol at the beginning of object data, (make sure it's referenced somehow?) and generate the actual MethodTable symbol as alt-entry.

You are right, this specific case may not be handled correctly. We would need to generate the symbol at the beginning of the data. We don't need to reference it, any reference to any alt_entry symbol is good enough to keep it alive (+ we still emit `N_NO_DEAD_STRIP`).

**Comment by filipnavara** [comment]
> We would need to generate the symbol at the beginning of the data.

Addressed in https://github.com/dotnet/runtime/pull/106224/commits/9ed567b108bb9f3b4ea4c7851e105f037330d975.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs [hunk 2]
```diff
@@ -414,24 +416,44 @@ private void EmitObject(string objectFilePath, IReadOnlyCollection<DependencyNod
...
+                // In the subsections-via-symbols model (Mach-O file format used by Apple) we need a symbol pointing
+                // to the beginning on the node without the `altEntry` flag. All symbols with non-zero offset are then
+                // produced with the `altEntry: true` flag. This makes the node appear as unbreakable atom for the
+                // native linker with individual symbol labels pointing into it.
+                //
+                // If the defined symbols of the node don't cover the zero offset we have to synthesize the initial
+                // label.
+                if (!haveStartSymbol && _usesSubsectionsViaSymbols)
+                {
+                    string startSymbol = currentSymbolName != null ? $"__start_{currentSymbolName}" :$"__start_{syntheticLabelCounter++}";
```
Comment (filipnavara) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:450]: I'd appreciate suggestions for the name here. I just chose something that didn't collide with existing symbols but I am not particularly happy about the name. I could try to emit emit it as local symbol too but that will require a bit more tweaking. Currently it's "private extern" (aka "hidden" in ELF terminology).

**Comment by MichalStrehovsky** [comment]
Correct me if I'm wrong:

1. If this is hydrated section:
Emit a synthetic symbol at the beginning of section
Emit all symbols within the section as altentry

2. If this is not a hydrated section:
If the symbol doesn't start at offset 0, emit a synthetic symbol at the beginning of objectdata packet. Emit everything else in the objectdata as altentry
If the symbol starts at offset 0, emit it as a normal symbol and everything else as altentry

So we now have two mechanisms to do the same thing. Is the motivation for introducing two mechanisms really only that if we were to do 1 with .data/.rdata sections, we might run into Apple linker bugs? Because I cannot imagine many bugs that would _not_ be hit by 1&2 above but would be hit if we do 1 with .data/.rdata sections. It feels like having both 1&2 is just introducing more codepaths that are involved and actually makes it more likely we'll run into a bug (or introduce a bug on our side).

**Comment by filipnavara** [comment]
> Correct me if I'm wrong:
> ...

Correct.

> So we now have two mechanisms to do the same thing.

It's a bit unfortunate. I would prefer if there was some implicit layout guarantee in the `ObjectData` model, ie. everything inside one node MUST be treated as unbreakable but CAN be relocated by the native linker. That guarantee currently holds for everything but the hydrated data. That's why I needed to special case it.

> Is the motivation for introducing two mechanisms really only that if we were to do 1 with .data/.rdata sections, we might run into Apple linker bugs?

There are multiple motivations.

It's not just that we can run into Apple linker bugs, we DO run into Apple linker bugs. The one-atom-per-section model in the `hydrated` data triggers a bug in the Apple linker. We are just lucky that `N_NO_DEAD_STRIP` cancels it out. I am not exactly sure yet of the full combination of conditions that are necessary to trigger/avoid the bug. I tried few obvious things without success.

Furthermore, the atom splitting algorithm is not driven just by the symbols. It's also driven by unwinding data for actual code. That makes the single-atom-per-section model inapplicable to code sections, or at very least more prone to hit other linker bugs.

Lastly, I still have open issue for COMDAT support and I don't want to make it harder to fix that in future. In Mach-O the implementation would likely amount to setting the `S_COALESCE` flag on the first symbol of the atom.

**Comment by MichalStrehovsky** [comment]
> Furthermore, the atom splitting algorithm is not driven just by the symbols. It's also driven by unwinding data for actual code. That makes the single-atom-per-section model inapplicable to code sections, or at very least more prone to hit other linker bugs.

I'm definitely not suggesting we do this for code sections. We don't have any assumptions about layout of code sections that I know of. The problems are only with data.

> Lastly, I still have open issue for COMDAT support and I don't want to make it harder to fix that in future. In Mach-O the implementation would likely amount to setting the S_COALESCE flag on the first symbol of the atom.

I think with #101969 we have very little motivation to do COMDAT folding. It would be meaningful only if we invest into multi-obj-compilation but we're not finding a compelling reason to do that. We might end up just deleting that mode. The benefits of COMDAT folding would be marginal at best outside of that mode. We can efficiently fold managed code and data in the object writer and there's little chance to deduplicate things outside of that.

**Comment by filipnavara** [comment]
> I'm definitely not suggesting we do this for code sections. We don't have any assumptions about layout of code sections that I know of. The problems are only with data.

That still means that we have to maintain a separate solution for code and data sections.

For code sections, at minimum we would need to bail out if someone uses a defining symbol with non-zero offset in a method node. If we allow it we will need to keep the code to mark such symbols as `N_ALT_ENTRY` and to produce the initial symbol for beginning of the node. I don't think there's any usage of this today. Correct me if I am wrong. Would it be okay to bake in this assumption (with asserts)?

For data sections, we can universally use the one-atom-per-section approach. I'm still bit uneasy about the broken behavior for the one-atom-per-section in the new Xcode linker though, so even if I find it compelling to switch all data sections to the single atom model I feel a bit uneasy about it until that issue is understood better. I'll try to spend a bit more time on it.

> I think with https://github.com/dotnet/runtime/pull/101969 we have very little motivation to do COMDAT folding.

That's fair. I think the code can be simplified quite a lot if we drop the COMDAT support. I didn't want to make it harder to implement.

**Comment by filipnavara** [comment]
Here's a rough draft of the alternative approach:
https://github.com/dotnet/runtime/compare/main...filipnavara:runtime:alt_entry_wip?expand=1

For code sections it keeps doing what it was doing, adds an assert.
For data sections it marks the local section symbol as `N_NO_DEAD_STRIP` and the rest of the symbols in the section as `N_ALT_ENTRY`.

Tested with Xcode 16 building the smoke tests with both the old and new linker.

On the upside, it produces a pattern that doesn't break the new linker for the data (by avoiding extra symbols at the same offset at the beginning of the section). On the downside, there still seems to be some dead-stripping issue with methods which requires keeping `N_NO_DEAD_STRIP`. I found a reasonable way to analyze it using trial-and-error and linker map files and I hope to resolve that eventually.

@MichalStrehovsky  Does this look more to your liking?

**Comment by filipnavara** [comment]
The new linker is turning out to be an extreme PITA. To ensure that I fixed all the dependency issues using `N_ALT_ENTRY` I removed `N_NO_DEAD_STRIP` from the symbols and recompiled the smoke tests (with the code in the [alt_entry_wip](https://github.com/dotnet/runtime/compare/main...filipnavara:runtime:alt_entry_wip?expand=1) branch linked in previous comment).

This was the result: 
```
# Dead Stripped Symbols:
# Address	Size    	File  Name
<<dead>>	0x00000070	[  4] _fram1_S_P_CoreLib_System_Exception__AppendExceptionStackFrame
<<dead>>	0x00000000	[  4] _fram2_S_P_CoreLib_System_Exception__AppendExceptionStackFrame
<<dead>>	0x0000006C	[  4] _fram1_S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
<<dead>>	0x00000000	[  4] _fram2_S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
<<dead>>	0x00000014	[  4] _fram1_S_P_CoreLib_System_RuntimeExceptionHelpers__RuntimeFailFast
```

This was the input:
```
000000000005fa00 (__TEXT,__managedcode) private external _S_P_CoreLib_System_Exception__AppendExceptionStackFrame
000000000005fbf0 (__TEXT,__managedcode) private external [alt entry] _fram1_S_P_CoreLib_System_Exception__AppendExceptionStackFrame
000000000005fc28 (__TEXT,__managedcode) private external [alt entry] _fram2_S_P_CoreLib_System_Exception__AppendExceptionStackFrame
...
00000000000602c0 (__TEXT,__managedcode) private external _S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
0000000000060684 (__TEXT,__managedcode) private external [alt entry] _fram1_S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
000000000006069c (__TEXT,__managedcode) private external [alt entry] _fram2_S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
...
0000000000060780 (__TEXT,__managedcode) private external _S_P_CoreLib_System_RuntimeExceptionHelpers__RuntimeFailFast
00000000000607bc (__TEXT,__managedcode) private external [alt entry] _fram1_S_P_CoreLib_System_RuntimeExceptionHelpers__RuntimeFailFast
```

There was many more `_fram[...]` "alt entry" entries that didn't get stripped, ale it works correctly with the old linker. Thankfully the `N_NO_DEAD_STRIP` workaround still works.

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky Does this look more to your liking?

Yes, this looks simpler and I like simple! Thanks!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs [hunk 3]
```diff
@@ -472,6 +473,10 @@ protected internal override void UpdateSectionAlignment(int sectionIndex, int al
 
             // Sort and insert all defined symbols
             var sortedDefinedSymbols = new List<MachSymbol>(definedSymbols.Count);
+            // We emit a local N_NO_DEAD_STRIP symbol for the beginning of all sections marked with
+            // KeepDataLayout to ensure the final layout is kept intact by the linker. This works in
+            // tandem with setting the N_ALT_ENTRY flag for all other symbols in the same section in
+            // the loop below.
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs:479]: Is the PR version still doing this?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs [hunk 4]
```diff
@@ -414,24 +415,33 @@ private void EmitObject(string objectFilePath, IReadOnlyCollection<DependencyNod
...
-                            global: true);
+                            global: true,
+                            altEntry: n.Offset != 0);
                     }
                 }
 
+                // Match the logic for KeepDataLayout in MachSection.CreateSection. If we are in a data section
+                // we don't need to have a symbol for beginning of the node data. For other nodes, particularly
+                // executable code, we enforce it though.
+                Debug.Assert(hasInitialEntrypoint || section.Type is SectionType.ReadOnly or SectionType.Writeable or SectionType.Uninitialized);
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:443]: Can we avoid duplicating the `section.Type is SectionType.ReadOnly or SectionType.Writeable or SectionType.Uninitialized` part?

**Review by MichalStrehovsky — COMMENTED** [review]
Looks good to me otherwise! Thank you!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs (hunk 3 above)
Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs:479]: Yes, it is.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs (hunk 4 above)
Comment (filipnavara) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs]: Done

**Comment by filipnavara** [comment]
> clang or linker seems to be segfaulting in the outerloop runs for at least one of the larger tests :(

I'll check it locally but likely not until tomorrow :-/

**Comment by filipnavara** [comment]
For posterity, the linke crash is stack overflow in the recursive call:
```
ld::tool::Resolver::markLive(ld::Atom const&, ld::tool::Resolver::WhyLiveBackChain*)
```

**Comment by agocke** [comment]
We've also reached out to Apple about this issue and hope they will get back to us

**Comment by filipnavara** [comment]
I will try to extract the non-problematic bits of this PR tomorrow into separate one(s).

I have some idea how to make the data section layout reliable without triggering the ld64 crash. In any case it’s going to be a non-trivial change and I need to test it properly.

**Comment by filipnavara** [comment]
> I have some idea how to make the data section layout reliable without triggering the ld64 crash. In any case it’s going to be a non-trivial change and I need to test it properly.

I came up with two hacky prototypes:

- Generate a single symbol for beginning of each data section and internally resolves the relocations pointing to the data section to (sectionSymbol + symbol value). The symbols in the data section are then not emitted at all. I had to do some nasty hacks to handle EHInfo & AssociatedData references from unwinding data, and globally visible data symbols (eg. `g_compilerEmbeddedSettingsBlob`, `g_compilerEmbeddedKnobsBlob`). It's not difficult to do all this and it would like be possible to make it a maintainable code change. I'm concerned about the impact on debuggability. I often relied on address to symbol lookup to identify method table from a pointer, and this would be lost. Perhaps there's a way to emit debug-only (`N_STABS`) or `L`-prefixed local symbols but that steers us back to the `N_ALT_ENTRY` solution which is prone to hit linker bugs.

- Get back to the state as of [9ed567b](https://github.com/dotnet/runtime/pull/106224/commits/9ed567b108bb9f3b4ea4c7851e105f037330d975) and try to clean it up so we don't do too many special cases. This means emitting the data in a linker visible way and potentially allowing it to reorder them. It also imposes additional restrictions:
  1. Symbol must be emitted at the beginning of each node (can be done by ObjWriter).
  2. All symbols with non-zero offsets need to be marked with `N_ALT_ENTRY` in order to keep the node data as single chunk (can be done by ObjWriter)
  3. Node data must be relocatable by the linker in the final output relative to other node data. This is currently violated by the contents of the `hydrated` section, hence the special casing I did earlier.

  The main obstacle to this solution would be remodeling how the `hydrated` data are emitted. The solution I used earlier doesn't really work well since it can still run into the limits of number of consecutive `N_ALT_ENTRY` symbols that crash the linker with stack overflow. We would likely need to take inspiration from the other prototype mentioned earlier and emit one symbol for the hydrated data and then use relative references to that symbol. And that once again leads to the debuggability degradation...

Neither of those solutions sounds like a slam dunk, but I wanted to share them anyway for posterity. Lastly, there's always the option of relying on the undefined behavior that the linker doesn't do the reordering for our code/data and that `N_NO_DEAD_STRIP` prevents any kind of dead code/data stripping. It may simply be good enough for now unless we come up with something cleaner.

I extracted the more obvious bits from this PR into #106442, #106444, and #106446. I'll close this PR since the general approach is a dead-end but I am open to continue discussion if there's any interest or guidance from Apple.

**Comment by filipnavara** [comment]
For posterity's sake, I received some feedback from Apple:

> The case new linker isn’t handling correctly is a combination of symbols at overlapping addresses and alt-entry symbols, like here:
>
> ```
> 00000000000602c0 (__TEXT,__managedcode) external _GetRuntimeException
> 00000000000602c0 (__TEXT,__managedcode) private external _S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
> 0000000000060684 (__TEXT,__managedcode) private external [alt entry] _fram1_S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException
> ```
>
> In this specific case marking `_S_P_CoreLib_System_RuntimeExceptionHelpers__GetRuntimeException` as an alt entry symbol would resolve the problem. That may not always work when using `-ld_classic` though (e.g., in the test_data.S sample) because old linker doesn’t handle alt-entry symbols at a start of a section. So if that can be a problem, we recommend using `-ld_classic` until the fix is available.

TL;DR:
- The new linker (ld-prime) has a bug with overlapping start symbols in combination with .alt_entry symbols. We can likely workaround this.
- `-ld_classic` continues to be a viable workaround even in Xcode 16, despite the warning that classic linker will be removed in future release (ie. future release is likely Xcode 17+). This removes some pressure to support the new linker for the .NET 9 timeframe (/cc @ivanpovazan).

I filed another issue (FB14897581) for the old linker (ld64) crashing with stack overflow for huge number of .alt_entry symbols and dead stripping. I asked for guidance on producing an unbreakable atom in DATA/BSS section with symbols that would be good enough not to break the "image lookup" experience in `lldb` debugger (to allow lookup of name from method table address).

**Comment by ivanpovazan** [comment]
Thanks for following up on this!

> `-ld_classic` continues to be a viable workaround even in Xcode 16 

Is there an official statement from Apple on this?

**Comment by filipnavara** [comment]
> Is there an official statement from Apple on this?

No official statement, as far as I know. The best I can tell is that I still keep receiving guidance `we recommend using -ld_classic until the fix is available` for bugs against Xcode 16 betas.
